replica/table_sstable_set.cc

```diff
@@ -0,0 +1,240 @@
+    virtual selector_and_schema_t make_incremental_selector() const override;
+
+    virtual flat_mutation_reader_v2 create_single_key_sstable_reader(
+            replica::column_family*,
+            schema_ptr,
+            reader_permit,
+            utils::estimated_histogram&,
+            const dht::partition_range&,
+            const query::partition_slice&,
+            tracing::trace_state_ptr,
+            streamed_mutation::forwarding,
+            mutation_reader::forwarding,
+            const sstable_predicate&) const override;
+
+    class incremental_selector;
+private:
+    // The for_each_sstable_set_* helpers guarantee atomicity
+    // only for each compaction group' compound_sstable_set,
+    // but not across compaction groups.
+    stop_iteration for_each_sstable_set_until(const dht::partition_range&, std::function<stop_iteration(lw_shared_ptr<sstable_set>)>) const;
+    future<stop_iteration> for_each_sstable_set_gently_until(const dht::partition_range&, std::function<future<stop_iteration>(lw_shared_ptr<sstable_set>)>) const;
+};
+
+stop_iteration table_sstable_set::for_each_sstable_set_until(const dht::partition_range& pr, std::function<stop_iteration(lw_shared_ptr<sstable_set>)> func) const {
+    if (auto scg = _table.single_compaction_group_if_available()) {
+        return func(scg->make_compound_sstable_set());
+    }
+    return _table.foreach_compaction_group_until(pr, [func = std::move(func)] (compaction_group& cg) {
+        return func(cg.make_compound_sstable_set());
+    });
+}
+
+future<stop_iteration> table_sstable_set::for_each_sstable_set_gently_until(const dht::partition_range& pr, std::function<future<stop_iteration>(lw_shared_ptr<sstable_set>)> func) const {
+    if (auto scg = _table.single_compaction_group_if_available()) {
+        return func(scg->make_compound_sstable_set());
+    }
+    return _table.foreach_compaction_group_gently_until(pr, [func = std::move(func)] (compaction_group& cg) {
+        return func(cg.make_compound_sstable_set());
+    });
+}
+
+std::vector<shared_sstable> table_sstable_set::select(const dht::partition_range& range) const {
+    std::vector<shared_sstable> ret;
+    for_each_sstable_set_until(range, [&] (lw_shared_ptr<sstable_set> set) {
+        auto ssts = set->select(range);
+        if (ret.empty()) {
+            ret = std::move(ssts);
+        } else {
+            if (ssts.size() > 1) {
+                ret.reserve(ret.size() + ssts.size());
+            }
+            std::move(ssts.begin(), ssts.end(), std::back_inserter(ret));
+        }
+        return stop_iteration::no;
+    });
+    tlogger.info("table_sstable_set::select: range={} ret={}", range, ret.size());
+    return ret;
+}
+
+lw_shared_ptr<const sstable_list> table_sstable_set::all() const {
+    auto ret = make_lw_shared<sstable_list>();
+    for_each_sstable_set_until(query::full_partition_range, [&] (lw_shared_ptr<sstable_set> set) {
```

Comment:
How efficient is this in skipping compaction groups we don't have?

Reply:
Not very efficient.
I think we could walk the intrusive list of compaction_group:s to pass through only the cg:s we have, skipping those we don't have.

Reply:
we already have this.